eventbus/jetstream/eventsource/jetstreamSourceConn.go

```diff
@@ -0,0 +1,41 @@
+package eventsource
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strconv"
+	"time"
+
+	eventbuscommon "github.com/argoproj/argo-events/eventbus/common"
+	jetstreambase "github.com/argoproj/argo-events/eventbus/jetstream/base"
+	nats "github.com/nats-io/nats.go"
+)
+
+type JetstreamSourceConn struct {
+	*jetstreambase.JetstreamConnection
+	eventSourceName string
+	randomGenerator *rand.Rand
+}
+
+func CreateJetstreamSourceConn(conn *jetstreambase.JetstreamConnection, eventSourceName string) *JetstreamSourceConn {
+	randSource := rand.NewSource(time.Now().UnixNano())
+	randomGenerator := rand.New(randSource)
+
+	return &JetstreamSourceConn{
+		conn, eventSourceName, randomGenerator,
+	}
+}
+
+func (jsc *JetstreamSourceConn) Publish(ctx context.Context,
+	evt eventbuscommon.Event,
+	message []byte) error {
+	// exactly once on the publishing side is done by assigning a "deduplication key" to the message
+	dedupKey := nats.MsgId(strconv.Itoa(jsc.randomGenerator.Intn(1000000)))
```

Comment:
Regarding: _"We have an auto generated UUID in the cloudevent ,and if the event source comes with any an ID (like Kafka) for each event, we use that ID."_ Are you saying that the code currently uses a Kafka, etc ID? I'm curious where that's happening, since I just see where it gets set randomly [here](https://github.com/argoproj/argo-events/blob/master/eventsources/eventing.go#L494).

Reply:
As far as the Message struct, I guess that would replace the current eventbuscommon.Event struct, essentially adding more members?

Reply:
https://github.com/argoproj/argo-events/blob/master/eventsources/eventing.go#L499-L504

Reply:
> As far as the Message struct, I guess that would replace the current eventbuscommon.Event struct, essentially adding more members?

Yeah, feeling a Message struct makes more sense. You decide.

Reply:
Okay, yeah, I like the idea of this having more members than the original Event struct.